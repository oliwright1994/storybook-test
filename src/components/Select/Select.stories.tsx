import createStory from '@bright/storybook/createStory'
import { boolean } from '@storybook/addon-knobs'
import React from 'react'

import Select from '.'

createStory('COMPONENTS | Select', [
  {
    label: 'standard',
    jsx: () => {
      const knobsGroups = ['Display', 'Elements']
      const isDisabled = boolean('Disabled', false, knobsGroups[0])
      const isInvalid = boolean('Invalid', false, knobsGroups[0])
      const isRequired = boolean('Required', false, knobsGroups[0])
      const isReadonly = boolean('Readonly', false, knobsGroups[0])
      const withLabel = boolean('With label', true, knobsGroups[1])
      const withDescription = boolean('With description', true, knobsGroups[1])
      const withErrorMessage = boolean('With error message', true, knobsGroups[1])

      return (
        <Select
          id="storybookSelect"
          label={withLabel ? 'Storybook select:' : ''}
          description={withDescription ? 'This is select description' : ''}
          errorMessage={withErrorMessage ? 'This is error message' : ''}
          isDisabled={isDisabled}
          isInvalid={isInvalid}
          isRequired={isRequired}
          isReadOnly={isReadonly}
        >
          <option value="first">First value</option>
          <option value="second" selected>
            Second value
          </option>
          <option value="third" disabled>
            Third value
          </option>
        </Select>
      )
    },
  },
])
