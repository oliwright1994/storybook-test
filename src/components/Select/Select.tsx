import Stack from '@bright/components/Stack'
import { Space } from '@bright/constants'
import {
  FormControl,
  FormErrorMessage,
  FormHelperText,
  FormLabel,
  Select as ChakraSelect,
  SelectProps as ChakraSelectProps,
} from '@chakra-ui/core'
import React, { FC } from 'react'

export interface SelectProps
  extends Pick<ChakraSelectProps, 'onChange' | 'isDisabled' | 'isInvalid' | 'isRequired' | 'isReadOnly'> {
  id: string
  label?: string
  description?: string
  errorMessage?: string
}

const Select: FC<SelectProps> = ({
  children,
  onChange,
  isDisabled,
  isInvalid,
  isRequired,
  isReadOnly,
  id,
  label,
  description,
  errorMessage,
}) => {
  return (
    <FormControl isInvalid={isInvalid} isRequired={isRequired}>
      <Stack space={Space.xxs}>
        {label && (
          <FormLabel htmlFor={id} padding={0} verticalAlign="baseline">
            {label}
          </FormLabel>
        )}
        {description && <FormHelperText margin={0}>{description}</FormHelperText>}
        <ChakraSelect
          onChange={onChange}
          isDisabled={isDisabled}
          isInvalid={isInvalid}
          isRequired={isRequired}
          isReadOnly={isReadOnly}
          id={id}
        >
          {children}
        </ChakraSelect>
        {errorMessage && <FormErrorMessage>{errorMessage}</FormErrorMessage>}
      </Stack>
    </FormControl>
  )
}

export default Select
