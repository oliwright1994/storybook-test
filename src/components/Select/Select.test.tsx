import { fireEvent } from '@testing-library/react'
import React from 'react'
import { renderWithBrightTheme } from 'testing/helpers/themeHelper'

import Select from '.'

describe('<Select />', () => {
  const defaultOptions = ['first', 'second']
  const defaultId = 'testSelect'
  const generateDefaultOptions = (): JSX.Element[] => defaultOptions.map(value => <option>{value}</option>)
  const label = 'This is label'
  const description = 'This is label'
  const errorMessage = 'This is error message'

  it('should render as expected', () => {
    const { asFragment } = renderWithBrightTheme(<Select id={defaultId}>{generateDefaultOptions()}</Select>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('displays text', () => {
    const { getByText } = renderWithBrightTheme(<Select id={defaultId}>{generateDefaultOptions()}</Select>)
    expect(getByText(new RegExp(defaultOptions[0], 'i'))).toBeTruthy()
  })

  it('displays label text', () => {
    const { getByText } = renderWithBrightTheme(
      <Select id={defaultId} label={label}>
        {generateDefaultOptions()}
      </Select>
    )
    expect(getByText(new RegExp(label, 'i'))).toBeTruthy()
  })

  it('displays description text', () => {
    const { getByText } = renderWithBrightTheme(
      <Select id={defaultId} description={description}>
        {generateDefaultOptions()}
      </Select>
    )
    expect(getByText(new RegExp(description, 'i'))).toBeTruthy()
  })

  it('displays error text', () => {
    const { getByText } = renderWithBrightTheme(
      <Select id={defaultId} errorMessage={errorMessage} isInvalid={!!errorMessage}>
        {generateDefaultOptions()}
      </Select>
    )
    expect(getByText(new RegExp(errorMessage, 'i'))).toBeTruthy()
  })

  it('fires onChange={} ', () => {
    const handleChange = jest.fn()
    const { getByLabelText } = renderWithBrightTheme(
      <Select id={defaultId} onChange={handleChange} label={label}>
        {generateDefaultOptions()}
      </Select>
    )
    const selectElement = getByLabelText(label)
    expect(handleChange).toHaveBeenCalledTimes(0)
    fireEvent.change(selectElement, { target: { value: defaultOptions[1] } })
    expect(handleChange).toHaveBeenCalled()
    expect(handleChange).toHaveBeenCalledTimes(1)
  })
})
