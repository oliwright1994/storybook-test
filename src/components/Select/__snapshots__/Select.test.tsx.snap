// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render as expected 1`] = `
<DocumentFragment>
  <div
    class="css-0"
    role="group"
  >
    <div
      class="css-0"
    >
      <li
        class="css-1lxwves"
      />
      <li
        class="css-qzdso"
      />
      <li
        class="css-qzdso"
      >
        <div
          class="css-1jke4yk"
        >
          <select
            class="css-111rvec"
            id="testSelect"
          >
            <option>
              first
            </option>
            <option>
              second
            </option>
          </select>
          <div
            class="css-18l1y83"
          >
            <svg
              aria-hidden="true"
              class="css-12c4avn"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </li>
      <li
        class="css-qzdso"
      />
    </div>
  </div>
</DocumentFragment>
`;
